cpp/src/io/json/json_reader.cu

```diff
@@ -0,0 +1,687 @@
+ * limitations under the License.
+ */
+
+#include "json_reader.hpp"
+
+#include <cuda_runtime.h>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <vector>
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <thrust/device_ptr.h>
+#include <thrust/execution_policy.h>
+
+#include <thrust/host_vector.h>
+
+#include <nvstrings/NVStrings.h>
+
+#include "cudf.h"
+#include "utilities/cudf_utils.h"
+#include "utilities/error_utils.hpp"
+#include "utilities/type_dispatcher.hpp"
+
+#include "io/comp/io_uncomp.h"
+#include "rmm/rmm.h"
+#include "rmm/thrust_rmm_allocator.h"
+
+#include "io/utilities/parsing_utils.cuh"
+#include "io/utilities/wrapper_utils.hpp"
+
+using string_pair = std::pair<const char *, size_t>;
+
+/**---------------------------------------------------------------------------*
+ * @brief Reads JSON-structured data and returns an array of gdf_columns.
+ *
+ * @param[in,out] args Structure containing input and output args
+ *
+ * @return gdf_error GDF_SUCCESS if successful, otherwise an error code.
+ *---------------------------------------------------------------------------**/
+gdf_error read_json(json_read_arg *args) {
+  // Check if the passed arguments are valid
+  CUDF_EXPECTS(args != nullptr, "The args parameter cannot be null.\n");
+  CUDF_EXPECTS(args->source != nullptr, "Input source cannot be null.\n");
+
+  JsonReader reader(args);
+
+  reader.parse();
+
+  reader.setOutputArguments(args);
+
+  return GDF_SUCCESS;
+}
+
+JsonReader::JsonReader(json_read_arg *args) : args_(args) {
+  // Check if the passed arguments are supported
+  CUDF_EXPECTS(args_->lines, "Only Json Lines format is currently supported.\n");
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief Estimates the maximum expected length or a row, based on the number
+ * of columns
+ *
+ * If the number of columns is not available, it will return a value large
+ * enough for most use cases
+ *
+ * @param[in] num_columns Number of columns in the JSON file (optional)
+ *
+ * @return Estimated maximum size of a row, in bytes
+ *---------------------------------------------------------------------------**/
+constexpr size_t calculateMaxRowSize(int num_columns = 0) noexcept {
+  constexpr size_t max_row_bytes = 16 * 1024; // 16KB
+  constexpr size_t column_bytes = 64;
+  constexpr size_t base_padding = 1024; // 1KB
+  if (num_columns == 0) {
+    // Use flat size if the number of columns is not known
+    return max_row_bytes;
+  } else {
+    // Expand the size based on the number of columns, if available
+    return base_padding + num_columns * column_bytes;
+  }
+}
+
+/*
+ * @brief Convert dtype strings into gdf_dtype enum;
+ *
+ * Returns GDF_invalid if the input string is not a valid dtype string
+ */
+gdf_dtype convertStringToDtype(const std::string &dtype) {
```

Comment:
@j-ieong I was considering it, but I'm not sure if it's guaranteed that we will use the same mapping for different APIs. If you think this will always be the case, I'll move this function into a common file.

Reply:
For JSON and CSV, it'll probably be similar. Parquet and ORC currently map their own types internally.

Reply:
Done, moved to a new common file.